os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -137,117 +141,157 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
 	close(fd);
 
 	return OK;
+
 errout_with_fd:
 	close(fd);
-
+	if (need_unlink) {
+		unlink(path);
+	}
 	return ERROR;
 }
 
 /****************************************************************************
- * Name: binary_manager_load_binary
+ * Name: binary_manager_exec_binary
  *
  * Description:
- *	 This function loads binary with index in binary table.
+ *	 This function loads and executes binary.
  *
  ****************************************************************************/
+static int binary_manager_exec_binary(int bin_idx, char *path, binary_header_t *header_data)
+{
+	int ret;
+	int retry_count;
+	load_attr_t load_attr;
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	struct binary_s *binp;
+#endif
+
+	strncpy(load_attr.bin_name, header_data->bin_name, BIN_NAME_MAX);
+	load_attr.bin_size = header_data->bin_size;
+	load_attr.compression_type = header_data->compression_type;
+	load_attr.ram_size = header_data->bin_ramsize;
+	load_attr.stack_size = header_data->bin_stacksize;
+	load_attr.priority = header_data->bin_priority;
+	load_attr.offset = CHECKSUM_SIZE + header_data->header_size;
+	strncpy(load_attr.bin_ver, header_data->bin_ver, BIN_VER_MAX);
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	binp = BIN_LOADINFO(bin_idx);
+	/* If current version is different from version to be loaded, Clean up backup data */
+	if (binp && strncmp(BIN_VER(bin_idx), header_data->bin_ver, strlen(BIN_VER(bin_idx)) + 1)) {
+		binp->reload = false;
+		binfmt_exit(binp);
+		binp = NULL;
```

Comment:
If we cleanup the binp, we don't need to set binp->reload to false.

Reply:
It is checked and memory is freed in binfmt_exit.